File: packages/canvas-py/src/langchain_canvas/protocol/events.py
"""Canvas Wire Protocol v1 — event envelopes.

Mirror of `packages/canvas-react/src/protocol/events.ts`. Every object streamed
to the client is one of these, discriminated by `type`. See `docs/02-protocol.md`
for the full specification.

Two families travel on the same wire:

* **chat** (`message.*`, `tool.*`)  — drives the conversation transcript.
* **canvas** (`canvas.*`)           — drives the canvas panel.

Plus a small **control** family (`error`, `done`).

These models are the *only* place the wire shape is defined on the backend. The
emitter builds them; the SSE bridge serializes them with `to_sse()`.
"""

from __future__ import annotations

import json
from typing import Any, Literal, Union

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

from .artifacts import Artifact, ArtifactStatus


class _Event(BaseModel):
    """Base for every wire event.

    Fields are declared in idiomatic snake_case but serialized to camelCase
    (`message_id` -> `messageId`) so the wire matches the TypeScript contract.
    `to_sse()` renders one SSE `data:` frame.
    """

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    def to_sse(self) -> str:
        payload = self.model_dump(by_alias=True, exclude_none=True)
        return f"data: {json.dumps(payload, ensure_ascii=False)}\n\n"


# --- chat family ----------------------------------------------------------------


class MessageDelta(_Event):
    type: Literal["message.delta"] = "message.delta"
    message_id: str
    text: str


class MessageEnd(_Event):
    type: Literal["message.end"] = "message.end"
    message_id: str


class ToolStart(_Event):
    type: Literal["tool.start"] = "tool.start"
    tool_call_id: str
    name: str


class ToolEnd(_Event):
    type: Literal["tool.end"] = "tool.end"
    tool_call_id: str
    ok: bool = True


# --- canvas family --------------------------------------------------------------


class CanvasCreate(_Event):
    type: Literal["canvas.create"] = "canvas.create"
    artifact: Artifact


class CanvasAppend(_Event):
    """Append `text` to the string at `data.<path>` (e.g. a document body)."""

    type: Literal["canvas.append"] = "canvas.append"
    id: str
    path: str
    text: str


class CanvasPatch(_Event):
    """JSON-merge-patch (RFC 7386) `patch` into the artifact's `data`."""

    type: Literal["canvas.patch"] = "canvas.patch"
    id: str
    patch: dict[str, Any]


class CanvasNodePatch(_Event):
    """Replace one element (by its `data-cid` path) in an html artifact.

    An O(1) surgical edit: the client resolves `cid` against the source HTML and
    swaps that element's outer HTML for `html`, instead of resending the page.
    """

    type: Literal["canvas.node_patch"] = "canvas.node_patch"
    id: str
    cid: str
    html: str


class CanvasReplace(_Event):
    """Replace the artifact wholesale — the client snapshots a new version."""

    type: Literal["canvas.replace"] = "canvas.replace"
    id: str
    artifact: Artifact


class CanvasStatus(_Event):
    type: Literal["canvas.status"] = "canvas.status"
    id: str
    status: ArtifactStatus


class CanvasClose(_Event):
    type: Literal["canvas.close"] = "canvas.close"
    id: str


# --- control family -------------------------------------------------------------


class ErrorEvent(_Event):
    type: Literal["error"] = "error"
    message: str


class DoneEvent(_Event):
    type: Literal["done"] = "done"


CanvasEvent = Union[
    CanvasCreate,
    CanvasAppend,
    CanvasPatch,
    CanvasNodePatch,
    CanvasReplace,
    CanvasStatus,
    CanvasClose,
]

StreamEvent = Union[
    MessageDelta,
    MessageEnd,
    ToolStart,
    ToolEnd,
    CanvasEvent,
    ErrorEvent,
    DoneEvent,
]

File: packages/canvas-react/src/protocol/events.ts
/**
 * Canvas Wire Protocol v1 — event envelopes. Mirror of
 * `langchain_canvas/protocol/events.py`. Every SSE frame is one `StreamEvent`,
 * discriminated by `type`. See `docs/02-protocol.md` for the specification.
 */

import type { Artifact, ArtifactStatus } from "./artifacts";

// --- chat family (drives the transcript) ---------------------------------------

export interface MessageDelta {
  type: "message.delta";
  messageId: string;
  text: string;
}

export interface MessageEnd {
  type: "message.end";
  messageId: string;
}

export interface ToolStart {
  type: "tool.start";
  toolCallId: string;
  name: string;
}

export interface ToolEnd {
  type: "tool.end";
  toolCallId: string;
  ok: boolean;
}

// --- canvas family (drives the panel) ------------------------------------------

export interface CanvasCreate {
  type: "canvas.create";
  artifact: Artifact;
}

/** Append `text` to the string at `data.<path>` (e.g. a document body). */
export interface CanvasAppend {
  type: "canvas.append";
  id: string;
  path: string;
  text: string;
}

/** JSON-merge-patch (RFC 7386) `patch` into the artifact's `data`. */
export interface CanvasPatch {
  type: "canvas.patch";
  id: string;
  patch: Record<string, unknown>;
}

/**
 * Replace a single element (by its `data-cid` tree path) inside an `html`
 * artifact with new outer HTML — an O(1) surgical edit that avoids resending the
 * whole page. The reconciler resolves the `cid` path against the source HTML.
 *
 * `slideId`/`nodeId` are set for deck artifacts, where the durable edit
 * address is `(deckId, slideId, nodeId)` — `cid` stays screen-only.
 */
export interface CanvasNodePatch {
  type: "canvas.node_patch";
  id: string;
  cid: string;
  html: string;
  slideId?: string;
  nodeId?: string;
}

/**
 * Report one slide's pipeline stage during deck conversion (extract ->
 * generate -> verify), so the client can show progress and let completed
 * slides be edited without waiting on the rest of the deck.
 */
export interface SlideStatus {
  type: "canvas.slide_status";
  id: string;
  slideId: string;
  stage: "extracting" | "generating" | "verifying" | "complete" | "degraded";
  detail?: string;
}

/**
 * Replace one slide's template HTML in a deck artifact — transmits a single
 * slide over the wire instead of resending the whole deck through
 * `canvas.patch`.
 */
export interface CanvasSlidePatch {
  type: "canvas.slide_patch";
  id: string;
  slideId: string;
  templateHtml: string;
}

/** Replace wholesale — the reconciler snapshots a new version. */
export interface CanvasReplace {
  type: "canvas.replace";
  id: string;
  artifact: Artifact;
}

export interface CanvasStatus {
  type: "canvas.status";
  id: string;
  status: ArtifactStatus;
}

/**
 * Promote the artifact's current state to a described version snapshot.
 * `revision` is the opaque store revision when a CanvasStore backs the canvas.
 */
export interface CanvasCommit {
  type: "canvas.commit";
  id: string;
  description: string;
  revision?: string;
  /** Set when this commit continues the version already on the rail: the
   *  entry is replaced rather than joined, so a burst of small saves reads
   *  as one work unit. */
  amends?: string;
}

// --- control family ------------------------------------------------------------

export interface ErrorEvent {
  type: "error";
  message: string;
}

export interface DoneEvent {
  type: "done";
}

// --- unions --------------------------------------------------------------------

export type ChatEvent = MessageDelta | MessageEnd | ToolStart | ToolEnd;

export type CanvasEvent =
  | CanvasCreate
  | CanvasAppend
  | CanvasPatch
  | CanvasNodePatch
  | SlideStatus
  | CanvasSlidePatch
  | CanvasReplace
  | CanvasStatus
  | CanvasCommit;

export type StreamEvent = ChatEvent | CanvasEvent | ErrorEvent | DoneEvent;

/** Narrow a `StreamEvent` to the canvas family. */
export function isCanvasEvent(event: StreamEvent): event is CanvasEvent {
  return event.type.startsWith("canvas.");
}

/** Narrow a `StreamEvent` to the chat family. */
export function isChatEvent(event: StreamEvent): event is ChatEvent {
  return event.type.startsWith("message.") || event.type.startsWith("tool.");
}
